src/backend/distributed/utils/background_jobs.c

```diff
@@ -0,0 +1,1325 @@
+								 "delay", secs)));
+
+			backgroundWorkerFailedStartTime = 0;
+		}
+
+
+		pid_t pid = 0;
+		GetBackgroundWorkerPid(handle, &pid);
+
+		ereport(LOG, (errmsg("found task with jobid/taskid: %ld/%ld",
+							 task->jobid, task->taskid)));
+
+		StartTransactionCommand();
+		PushActiveSnapshot(GetTransactionSnapshot());
+
+		/*
+		 * Reload task while holding a new ExclusiveLock on the table. A separate process
+		 * could have cancelled or removed the task by now, they would not see the pid and
+		 * status update, so it is our responsibility to stop the backend and _not_ write
+		 * the pid and running status.
+		 *
+		 * The lock will release on transaction commit.
+		 */
+		LockRelationOid(DistBackgroundTaskRelationId(), ExclusiveLock);
+
+		oldContext = MemoryContextSwitchTo(perTaskContext);
+		task = GetBackgroundTaskByTaskId(task->taskid);
+		MemoryContextSwitchTo(oldContext);
+
+		if (!task || task->status == BACKGROUND_TASK_STATUS_CANCELLING ||
+			task->status == BACKGROUND_TASK_STATUS_CANCELLED)
+		{
+			task->status = BACKGROUND_TASK_STATUS_CANCELLED;
+			UpdateBackgroundTask(task);
+			UpdateBackgroundJob(task->jobid);
+
+			PopActiveSnapshot();
+			CommitTransactionCommand();
+
+			/*
+			 * Terminate backend and release shared memory to not leak these resources
+			 * across iterations.
+			 */
+			TerminateBackgroundWorker(handle);
+			dsm_detach(seg);
+
+			/* there could be an other task ready to run, let a new loop decide */
+			continue;
+		}
+
+		/*
+		 * Now that we have verified the task has not been cancelled and still exist we
+		 * update it to reflect the new state
+		 */
+		task->status = BACKGROUND_TASK_STATUS_RUNNING;
+		SET_NULLABLE_FIELD(task, pid, pid);
+
+		/* Update task status to indicate it is running */
+		UpdateBackgroundTask(task);
+		UpdateBackgroundJob(task->jobid);
+
+		PopActiveSnapshot();
+		CommitTransactionCommand();
+
+		MemoryContextSwitchTo(perTaskContext);
+
+		bool hadError = false;
+		StringInfoData message = { 0 };
+		initStringInfo(&message);
+
+		while (GetBackgroundWorkerPid(handle, &pid) != BGWH_STOPPED)
+		{
+			int latchFlags = WL_LATCH_SET | WL_TIMEOUT | WL_EXIT_ON_PM_DEATH;
+			int rc = WaitLatch(MyLatch, latchFlags, (long) 1000, PG_WAIT_EXTENSION);
+
+			if (rc & WL_LATCH_SET)
+			{
+				ResetLatch(MyLatch);
+				CHECK_FOR_INTERRUPTS();
+			}
+
+			shm_toc *toc = shm_toc_attach(CITUS_BACKGROUND_TASK_MAGIC,
+										  dsm_segment_address(seg));
+			shm_mq *mq = shm_toc_lookup(toc, CITUS_BACKGROUND_TASK_KEY_QUEUE, false);
+			shm_mq_handle *responseq = shm_mq_attach(mq, seg, NULL);
+
+			ConsumeTaskWorkerOutput(responseq, &message, &hadError);
+
+			shm_mq_detach(responseq);
+		}
+
+		StartTransactionCommand();
+		PushActiveSnapshot(GetTransactionSnapshot());
+
+		/*
+		 * There is a small race condition between the last loop and the backend being
+		 * terminated. We re-consume the output of the process here again in case we have
+		 * missed a message.
```

Comment:
if the process writes a lot of messages and then dies, are we guaranteed to get all of them? (including the one that tells us what happened in the end?) I wonder whether it would be better to just call ConsumeTaskWorkerOutput until we get SHM_MQ_DETACHED.

Reply:
I think it makes a lot of sense to consume everything till we reach `SHM_MQ_DETACHED`, this will require some refactoring of the code though, so let me tinker a bit with that later tonight.

This might be the very last thing I change on this PR once it is the only outstanding item.

Reply:
agree, this is the last outstanding item

Reply:
> I wonder whether it would be better to just call ConsumeTaskWorkerOutput until we get SHM_MQ_DETACHED.

I have been looking into this, and it seems that we already do.

As you have noted above, we read from the shm_mq with `nowait` set to `false`. This makes that the call to read from the queue blocks. Causing some problems when we want to implement parallel workers.

However, due to the blocking behaviour and the loop we have untill we receive a response from our reader that is _not_ `SHM_MQ_SUCCESS` we will only ever exit the loop when we receive a `SHM_MQ_DETACHED`. There is only one other response we could get, which is `SHM_MQ_WOULD_BLOCK` which is returned only when `nowait` is `true`.

That said, I think the race I was anticipating here already assumed that we would be reading in a non-blocking manner.

We could do two things:
 - in the loop above we could read with `nowait` set to `true`. And finalize the consumption of the message queue here in a blocking manner like we currently do to assure we reach the `SHM_MQ_DETACHED` state.
 - leave the code as is, albeit, we probably want to change the comment above with the new insights.

Reply:
if ConsumeTaskWorkerOutput already runs to completion, do we still need the second ConsumeTaskWorkerOutput call below?

Reply:
actually, the race condition is that the process can complete before `while (GetBackgroundWorkerPid(handle, &pid) != BGWH_STOPPED)`

I think we can just remove that loop and call it a day?